### PR TITLE
Initial PR outlining the governance for the project

### PR DESCRIPTION
**Pull request by madolson**
Initial PR to add a governance doc outlining permissions for the main Valkey project as well as define responsibilities for sub-projects.

### CODE REVIEW

**Comment by natoscott** [comment]
@madolson LGTM FWIW :)  BTW, there's a typo 'Ericcson' -> 'Ericsson'

**Review by enjoy-binbin — APPROVED** [review]

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]: "New" seems redundant.

```suggestion
Maintainers (and accordingly, TSC members) may be added or removed by no less than 2/3 affirmative vote of the current TSC.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]: Does "maintainers" have full-commit access to **all** repositorie or just the main repository?

if it is the later, do we want to qualify "maintainers" and introduce different sub "maintainers" group? like doc/client/website/etc?

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md:18]: what is the definition of "members who are not maintainers"? committers? What is the process of removing committers?

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]: the delegation is likely a "major-decision" as well. also delegation can be revoked, at the discretion of the TSC.

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
The TSC will give at least two weeks for all members to submit their vote.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]: Should we state explicitly that this includes all **future** repositories under the Valkey-io org as well? or any new repo will automatically be subject to this governance?

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]: Can a TSC member call for a vote as well or this responsibility lies with the chair only?

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md:53]: can a non-TSC member request the TSC to take a vote on a TSC member's membership? if yes, under what situation can this be requested? what would a (formal) process look like?

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]: >The discussion will be open to the public

Should we talk about the form of "public disclosure" of these major decision discussions in this doc? or it will be clarified separately?

File: MAINTAINERS.md
Comment (hpatro) [inline MAINTAINERS.md:10]: Does affiliation has any significance to the project? Does an individual need to disclose any change in affiliation?

File: GOVERNANCE.md
Comment (hpatro) [inline GOVERNANCE.md:46]: What is the tie resolution?

**Comment by zuiderkwast** [comment]
Good job! Is there anything special with the filename `GOVERNANCE.md`?

Previously, this document was in the doc repo which means it became a page on the website. That's nice too.
https://github.com/valkey-io/valkey-doc/blob/0536b897b00f5d12057477719e3d2e82e5c63fcd/topics/governance.md (or https://web.archive.org/web/20210814154637/https://redis.io/topics/governance)

**Comment by madolson** [comment]
> Good job! Is there anything special with the filename GOVERNANCE.md?
> Previously, this document was in the doc repo which means it became a page on the website. That's nice too.
https://github.com/valkey-io/valkey-doc/blob/0536b897b00f5d12057477719e3d2e82e5c63fcd/topics/governance.md (or https://web.archive.org/web/20210814154637/https://redis.io/topics/governance)

My thought process was that since we already recursively copy Valkey for the website, we can simply render the GOVERNANCE.md doc on the website without it being in the doc. I have no strong opinion though.

File: MAINTAINERS.md
Comment (madolson) [reply MAINTAINERS.md:10]: I think it's useful for transparency. If you are unaffiliated, that's fine too.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md:46]: Not doing the thing I suppose. I don't know we want a tie breaking mechanism if it's 3-3 (or some other even/even in the future), I feel like we'll eventually come to a consensus.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md:18]: I use maintainer/committer interchangeably. I can clarify this is members of the TSC. For example, if we decide Jeff Bezos should be a TSC member, we can add him without adding him as a maintainer to the project if we want.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md]: I can emphasize this point, as well as add it above

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md:53]: I'm going to say no. If a non-TSC member can't get a TSC member to agree with them it's probably not worth forcing.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md]: I tried to be vague when I can be. Do you think we need more detail on what counts as public?

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md]: The wording here isn't great, the notion was that the TSC Chair would facilitate it if we needed it, but anyone can request it. I don't know if we need a chair though.

File: GOVERNANCE.md
Comment (PingXie) [reply GOVERNANCE.md:18]: Yeah let's define the roles precisely. for the record, I don't want Jeff on the TSC but fine with him being a maintainer - is it possible?

File: GOVERNANCE.md
Comment (PingXie) [reply GOVERNANCE.md]: maybe not in this doc but would be good to document it somewhere else. not a blocker for this PR.

File: GOVERNANCE.md
Comment (PingXie) [reply GOVERNANCE.md:46]: from the process' perspective, I think it is worth defining a tie-breaking process, just for the sake of completeness.  I think we talked about involving LF in the past. Would it make sense to add something along that line?

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md:18]: > I don't want Jeff on the TSC but fine with him being a maintainer - is it possible?

Yeah, although it's not clear how we would make that happen at the moment.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md]: Maintainers main the file the MAINTAINERS.md file is in. I see the lack of precision, will fix.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md:46]: I still think the answer is status quo wins by default. Like, if the questions is, should we add a feature or not, 3-3 vote means we don't do it.

**Comment by zuiderkwast** [comment]
> My thought process was that since we already recursively copy Valkey for the website, we can simply render the GOVERNANCE.md doc on the website without it being in the doc. I have no strong opinion though.

That may be the current solution for the website, but I don't think it's the best solution. The redis-io website didn't do this afaict, but then instead there's some generated content checked in to the doc repo instead. I think they used the script utils/generate-commands-json.py fetching info from a running node using to produce the JSON which was stored as commands.json in the doc repo, re-generated and updated manually. Probably the reply schemas were generated once, but later edited manually. Eveything else was from the doc repo IIUC.

The separation between doc repo and code repo is not very clear but I'd like to think that the doc repo should be useful on its own, for generating man pages or other things. It's under a different license as well (CC BY-SA 4.0).

Somehow I feel that a lot of documents don't belong in a code repo that you clone, build and install. README and license do and probably a few more. Are these "about" pages (governance, maintainers, code-of-conduct, etc.) part of documentation though? If they target contributors, do they belong in the code repo and if they target users, they are docs?

If we want to link to these from various pages in the docs, it's also a sign that they may be docs.

File: GOVERNANCE.md
Comment (PingXie) [reply GOVERNANCE.md:46]: "status quo wins" does not cover all cases. We could have a situation where the status quo is A but the split is between B and C. The process should have a "catch all" clause for the sake of completeness and default to "higher level" authority is a better process IMO.

File: GOVERNANCE.md
Comment (zuiderkwast) [reply GOVERNANCE.md]: "New" is wrong when it's about removing maintainers.

File: GOVERNANCE.md
Comment (PingXie) [reply GOVERNANCE.md:18]: I am fine with leaving this situation open for now. The reason I am asking is that I think we need to scale out better in the future. The TSC will not be able to review/approve every change say 2 years down the road. I also don't think we should expand TSC for the sake of "scaling". So one option for scaling is to have "maintainers" not on TSC. That said, not a blocker for me for this PR.

File: MAINTAINERS.md
Comment (hwware) [inline MAINTAINERS.md:15]: Can we list the maintainer sorted by Maintainer first character， OR GitHub ID, OR Affiliation? Now it looks like a little bit messy.

Comment (madolson) [reply MAINTAINERS.md:15]: Will do by github ID, which I will say is the most natural to lexicographically order by.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md:46]: > "status quo wins" does not cover all cases. We could have a situation where the status quo is A but the split is between B and C.

Then we do A. I don't think bringing in the LF makes sense to break ties. I have enough trust in us that if A is a bad choice, we will figure it out.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md]: I put this in notes for next meeting.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md:53]: I don't think this needs any change, but will leave unresolved if you still disagree.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md]: Someone suggested it's "practically" useful to have a chair, if for nothing else, they're the person you go talk to make something happen.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md]: Will add future for precision.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md:18]: My vision, at least, is the TSC should scale to like 8-12 or so people. I think that is a reasonable number for the core. We should further scale by  having folks be maintainers of modules and other sub-repositories like clients.

I'll add that the governance is written so we can solve that problem, later.

**Comment by madolson** [comment]
@soloestoy Would appreciate your feedback when you have a moment. I'm going to wait for explicit approval from 3/5 remaining maintainers before merging.

File: MAINTAINERS.md
Comment (zuiderkwast) [inline MAINTAINERS.md]: Correcting my name.

My github username is a funny spelling (like if it were Dutch, which it isn't).

```suggestion
| Viktor Söderqvist  | [zuiderkwast](https://github.com/zuiderkwast)   | Ericsson    |
```

File: GOVERNANCE.md
Comment (PingXie) [reply GOVERNANCE.md:18]: >I'll add that the governance is written so we can solve that problem, later.

SGTM. We can have that discussion when we cross that bridge.

File: GOVERNANCE.md
Comment (PingXie) [reply GOVERNANCE.md]: Yeah, having a chair makes sense. My question was more on clarifying the roles of the chair and members. Reading the clauses again, I think it works for me.

File: GOVERNANCE.md
Comment (PingXie) [reply GOVERNANCE.md:53]: agreed.

File: GOVERNANCE.md
Comment (PingXie) [reply GOVERNANCE.md:46]: SGTM

File: MAINTAINERS.md
Comment (enjoy-binbin) [inline MAINTAINERS.md]:
```suggestion
| Zhao Zhao           | [soloestoy](https://github.com/soloestoy)       | Alibaba     |
| Viktor Söderqvist   | [zuiderkwast](https://github.com/zuiderkwast)   | Ericsson    |
```

**Review by PingXie — APPROVED** [review]
LGTM. Thanks @madolson!

File: MAINTAINERS.md
Comment (soloestoy) [inline MAINTAINERS.md:7]: we can add a sentence here: ordered by github id :)

File: GOVERNANCE.md
Comment (soloestoy) [inline GOVERNANCE.md:6]: I'm a bit confused about the relationship between "maintainers" and "TSC". Although currently all TSC members are maintainers of the valkey core repo, but in the future these two groups may be changed separately. I think it would be better to describe them separately.

For example, we could use `MAINTAINERS.md` in the valkey repo to specifically detail the maintainers of the valkey repo. Additionally, introducing a `TSC.md` directly under valkey-io (maybe a new governance or TSC repo) to clarify the composition of the TSC, highlighting their role in overseeing all projects under the valkey-io. This setup ensures that any future additions of maintainers for other clients or modules to the TSC wouldn't impact the list of maintainers for the valkey repo.

File: GOVERNANCE.md
Comment (soloestoy) [inline GOVERNANCE.md:30]: I think it would also be beneficial to include examples of clients and modules, such as "adding a new client or introducing a new module".

File: GOVERNANCE.md
Comment (zuiderkwast) [inline GOVERNANCE.md:4]: "the Valkey-io organization" just means the github org. It sounds like were defining our project in terms of whats under the github org, but it's rather the other way around. Can we write it in a different way?

```suggestion
The Valkey project stores its repositories on GitHub under the valkey-io organization.
```

Comment (madolson) [reply GOVERNANCE.md:4]: I didn't want to tie the governance to specifically to github. Maybe I'll word it like:
```
The Valkey project includes all of the current and future repositories under the Valkey-io GitHub organization or whatever future code hosting platform the project uses.
```

?

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md:6]: At least for now, practically maintainers = TSC. I think it makes the most sense to amend the document when that needs to change.

File: GOVERNANCE.md
Comment (zuiderkwast) [reply GOVERNANCE.md:4]: Sure. Or just keep it as is. It doesn't matter much.

**Review by zuiderkwast — APPROVED** [review]

**Review by soloestoy — APPROVED** [review]

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
A vote shall also be called when an issue or pull request is marked as a major decision, which are decisions that have a significant impact on the Valkey architecture or design.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
* Unreachable Member: If a member is unresponsive for more than six months, the remaining active members of the TSC may vote to remove the unreachable member by a simple majority.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
The TSC may delegate decision making for other projects within the Valkey organization to the maintainers responsible for those projects.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
Projects within the Valkey organization must indicate the individuals with commit permissions by updating the MAINTAINERS.md within their repositories.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
The Valkey project is managed by a Technical Steering Committee (TSC) composed of the maintainers of the Valkey repository.
```

File: GOVERNANCE.md
Comment (PingXie) [reply GOVERNANCE.md:4]: is it "Valkey" or "Valkey-io" organization?

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
The TSC shall appoint a Chair responsible for organizing TSC meetings.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
The TSC is responsible for oversight of all technical, project, approval, and policy matters for Valkey.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
The TSC shall strive for all decisions to be made by consensus.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
Rather, the TSC shall determine consensus based on their good faith consideration of a number of factors, including the dominant view of the TSC and nature of support and objections.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
The TSC shall document evidence of consensus in accordance with these requirements.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
If consensus cannot be reached, the TSC shall make the decision by a vote.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
The discussion shall be open to the public, with the notable exception of discussions involving embargoed security issues or the addition or removal of maintainers, which will be private.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
The TSC shall give at least two weeks for all members to submit their vote.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
Delegation of decision making for a project is considered a major decision, and shall happen with an explicit vote.
```

File: GOVERNANCE.md
Comment (PingXie) [inline GOVERNANCE.md]:
```suggestion
Maintainers are defined as individuals with full commit access to a repository, which shall be in sync with the MAINTAINERS.md file in a given projects repository.
```

**Review by PingXie — COMMENTED** [review]
LGTM. some wording recommendation.

File: GOVERNANCE.md
Comment (madolson) [reply GOVERNANCE.md:4]: It's currently valkey-io organization. We're working on getting valkey though.

**Review by hwware — APPROVED** [review]
I would like to approve this hard work.

**Pull request merged**
